crates/services/p2p/src/config/connection_tracker.rs

```diff
@@ -0,0 +1,50 @@
+use super::{
+    fuel_authenticated::Approver,
+    peer_ids_set_from,
+};
+use crate::peer_manager::ConnectionState;
+use libp2p::{
+    Multiaddr,
+    PeerId,
+};
+use std::{
+    collections::HashSet,
+    sync::{
+        Arc,
+        RwLock,
+    },
+};
+
+/// A `ConnectionTracker` allows either Reserved Peers or other peers if there is an available slot.
+/// It is synced with `PeerManager` which keeps track of the `ConnectionState`.
+#[derive(Debug, Clone)]
+pub(crate) struct ConnectionTracker {
+    reserved_nodes: HashSet<PeerId>,
+    connection_state: Arc<RwLock<ConnectionState>>,
```

Comment:
Why do we want to track it on the transport level instead of behavior?

Maybe it would be simpler to handle it inside of `on_connection_handler_event` of `PeerManagerBehaviour`? In this case, we don't need any locks

Reply:
I see that we do it because of https://github.com/FuelLabs/fuel-core/pull/876#discussion_r1065400628. But does it really consume a slot between our check in the `FuelNodeAuthentication` and when we can process it here?
<img width="691" alt="image" src="https://user-images.githubusercontent.com/18346821/211938776-383fc425-a489-4f2b-be15-c1ba4ed2d3a3.png">

Reply:
In reserved nodes only mode, there's no good reason to allow connections from any non-reserved peers. Consider it as a DOS mitigation vector, because otherwise a sybil network could target validator nodes with lots of connections and interfere with the connection limit and potentially interfere with consensus / liveness.

For regular nodes that maintain connections with both reserved and unreserved peer ids, deferring the peer disconnection to an async event could still allow for DOS issues. For instance a sybil attacker could saturate the connections of a sentry node and occupy connections slots that should be used for the reserved node, causing it to desync.

By tracking this at the transport level, we minimize the number of resources an attacker can consume. Even in the normal mode of operation, waiting for connections to be established and then asynchronously dropping them later could cause scenarios like what I described above.

Reply:
Yes, I understand that. My question was, what exactly will we save by this change=)

I found an answer. We will skip [this](https://github.com/libp2p/rust-libp2p/blob/5782a96af2241f068ee0cb4a99503db5e2941035/swarm/src/connection/pool.rs#L622) step, For us it saves:
- We don't increase connection counters. So it fixes the problem described by you. We can always consider having unlimited connections, but it is a bad idea based on the points below=)
- We don't spawn a new `new_for_established_connection` task(the major saving) with loop inside to process data.
- We don't create a new `ConnectionEstablished` event - minus one additional iteration in the swarm.

So yeah, it makes sense.

Hmm, why the `substrate` doesn't do that=) They process peers in the behavior and allow `10000` connections.

Reply:
Maybe a good question for tomaka :)

Reply:
If I had to guess, I could see how setting a very high connection threshold and then maintaining a smaller set of peers in network behavior may help with things like peer discovery if the bootstrap nodes are saturated. In that case it would only make sense to have transport auth for nodes that are configured as reserved-only. I would have to take a look to see if allowing more connections than peers would be enough to allow for peer discovery though.